### PR TITLE
Rewrite tx-pool

### PR DESCRIPTION
**Pull request by zhangsoledad**
The existing tx-pool code has many potential issues, the PR focus those issue fix.
* Make true tx callback covered in the right place，use callback achieve statics 
   https://github.com/nervosnetwork/ckb/pull/2634/commits/f45958d2229612776950a2dbf3c6482d607ee5f8 https://github.com/nervosnetwork/ckb/pull/2634/commits/83f148500ef2f1e15433d7f6acddc8b01e720f04 https://github.com/nervosnetwork/ckb/pull/2634/commits/02512ed7f134dd92d6e2a5fcff1885362b5aab7b
* short-circuit resolve transaction https://github.com/nervosnetwork/ckb/pull/2634/commits/ef1996277cf65bdacdafac96fbdcc24177073823
* tx is relay by batch, if some tx is rejected (maybe due to the transaction fee threshold of different configurations), the whole batch of transactions should not be discarded https://github.com/nervosnetwork/ckb/pull/2634/commits/e615dc0eee7d35301e89de68d428b5398f60c70f
* cache resolved tx, avoid duplication resolve https://github.com/nervosnetwork/ckb/pull/2634/commits/98d16f8441689f767dad41c6f96013176b10813c
* make verify remote tx use the declared cycles as the limit https://github.com/nervosnetwork/ckb/pull/2634/commits/8a25a4ff2033751c01285cd5baf30e0bf0910b87
* rewrite tx relay process https://github.com/nervosnetwork/ckb/pull/2634/commits/99ecabb25b839fc1d14d1fb9710c97f8bf66bcb3
* rewrite orphan tx process https://github.com/nervosnetwork/ckb/pull/2634/commits/e0d54c0c0cf41efeb82367996983e60deeda9b81

### CODE REVIEW

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 480
  - CKB Version: e0d54c0c0cf41efeb82367996983e60deeda9b81
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Review by yangby-cryptape — APPROVED** [review]
It looks good to me.

Only a few issues about coding style in my opinion, and I think we can do optimization later, so that we can merge this big PR asap.

File: tx-pool/src/component/entry.rs [hunk 1]
```diff
@@ -61,32 +27,46 @@ pub struct TxEntry {
     pub ancestors_cycles: Cycle,
     /// ancestors txs count
     pub ancestors_count: usize,
-    /// related out points (cell deps includes cell group itself)
-    pub related_out_points: Vec<OutPoint>,
+    // /// related out points (cell deps includes cell group itself)
```
Comment (quake) [inline tx-pool/src/component/entry.rs:30]: Please remove commented out code

File: tx-pool/src/component/orphan.rs [hunk 2]
```diff
@@ -1,325 +1,137 @@
...
                     None
                 }
             })
             .cloned()
             .collect();
-
-        for id in ids {
-            self.recursion_remove(&id);
+        for id in expires {
+            evicted += 1;
```
Comment (quake) [inline tx-pool/src/component/orphan.rs:100]:
```suggestion
        let mut evicted = expires.len();
        for id in expires {
```

File: tx-pool/src/component/orphan.rs [hunk 3]
```diff
@@ -1,325 +1,137 @@
...
-                for outpoint in entry.transaction.output_pts() {
-                    if let Some(ids) = self.edges.remove(&outpoint) {
-                        queue.extend(ids);
-                    }
-                }
-            }
+        while self.len() > DEFAULT_MAX_ORPHAN_TRANSACTIONS {
+            evicted += 1;
+            // Evict a random orphan:
+            let id = self.entries.keys().cloned().next().expect("bound checked");
```
Comment (quake) [inline tx-pool/src/component/orphan.rs:107]:
```suggestion
            let id = self.entries.keys().next().cloned().expect("bound checked");
```

File: tx-pool/src/component/orphan.rs [hunk 4]
```diff
@@ -1,325 +1,137 @@
...
-        self.vertices.insert(short_id, entry)
+        self.shrink_to_fit();
     }
 
-    fn prune(&mut self) {
-        if self.size() < self.prune_threshold {
-            return;
-        }
-
+    pub fn limit_size(&mut self) -> u64 {
```
Comment (quake) [inline tx-pool/src/component/orphan.rs:84]:
```suggestion
    pub fn limit_size(&mut self) -> usize {
```

File: tx-pool/src/pool.rs [hunk 5]
```diff
@@ -226,58 +199,43 @@ impl TxPool {
...
-            || self.orphan.contains_key(id)
+        self.pending.contains_key(id) || self.proposed.contains_key(id)
     }
 
     /// Returns tx with cycles corresponding to the id.
     pub fn get_tx_with_cycles(
         &self,
         id: &ProposalShortId,
-    ) -> Option<(TransactionView, Option<Cycle>)> {
+    ) -> Option<(&TransactionView, Option<Cycle>)> {
```
Comment (quake) [inline tx-pool/src/pool.rs:209]: I think the `Cycle` cannot be `None` after refactoring, we can change the return type to:

```suggestion
    ) -> Option<(&TransactionView, Cycle)> {
```

File: tx-pool/src/process.rs [hunk 6]
```diff
@@ -392,104 +370,217 @@ impl TxPoolService {
...
-            }
+    pub(crate) fn broadcast_tx(&self, origin: PeerIndex, tx_hash: Byte32) {
+        if let Err(e) = self.tx_relay_sender.send((origin, tx_hash)) {
+            error!("tx-pool broadcast_tx internal error {}", e);
         }
-        Ok(())
     }
 
-    pub(crate) async fn process_txs(
+    #[allow(unused_variables)]
```
Comment (quake) [inline tx-pool/src/process.rs:528]: unnecessary clippy hint?

**Review by quake — CHANGES_REQUESTED** [review]

File: benches/benches/benchmarks/overall.rs [hunk 7]
```diff
@@ -46,6 +45,43 @@ fn block_assembler_config() -> BlockAssemblerConfig {
...
+        ping_interval_secs: 15,
+        ping_timeout_secs: 20,
+        connect_outbound_interval_secs: 1,
+        discovery_local_address: true,
+        upnp: false,
+        bootnode_mode: true,
+        max_send_buffer: None,
+        sync: SyncConfig::default(),
+        reuse: true,
+    };
```
Comment (yangby-cryptape) [inline benches/benches/benchmarks/overall.rs:69]:
```suggestion
    let config = NetworkConfig {
        max_peers: 19,
        max_outbound_peers: 5,
        path: tmp_dir.path().to_path_buf(),
        ping_interval_secs: 15,
        ping_timeout_secs: 20,
        connect_outbound_interval_secs: 1,
        discovery_local_address: true,
        bootnode_mode: true,
        reuse: true,
        ..Default::default() 
    };
```

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 486
  - CKB Version: 846021dacd2058c6b40ae1b3b94e7854f1290305
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Review by quake — APPROVED** [review]

**Review by yangby-cryptape — APPROVED** [review]

**Comment by yangby-cryptape** [comment]
bors r=quake,yangby-cryptape

**Pull request merged**